plugins/faucet/faucet.go

```diff
@@ -0,0 +1,344 @@
+package faucet
+
+import (
+	"context"
+	"time"
+
+	"github.com/cockroachdb/errors"
+	"github.com/iotaledger/goshimmer/client/wallet/packages/seed"
+	"github.com/iotaledger/goshimmer/packages/app/faucet"
+	"github.com/iotaledger/goshimmer/packages/core/ledger"
+	"github.com/iotaledger/goshimmer/packages/core/ledger/utxo"
+	"github.com/iotaledger/goshimmer/packages/core/ledger/vm/devnetvm"
+	"github.com/iotaledger/goshimmer/packages/core/ledger/vm/devnetvm/indexer"
+	"github.com/iotaledger/goshimmer/packages/core/tangleold"
+	"github.com/iotaledger/goshimmer/packages/node/clock"
+	"github.com/iotaledger/goshimmer/plugins/blocklayer"
+	"github.com/iotaledger/hive.go/core/crypto/ed25519"
+	"github.com/iotaledger/hive.go/core/generics/event"
+	"github.com/iotaledger/hive.go/core/generics/lo"
+	"github.com/iotaledger/hive.go/core/generics/queue"
+	"github.com/iotaledger/hive.go/core/identity"
+	"github.com/iotaledger/hive.go/core/types"
+)
+
+var (
+	numOfOutputsToSplit  = 100
+	maxTxBookedAwaitTime = 5 * time.Second
+	waitForAcceptance    = 10 * time.Second
+	maxWaitAttempts      = 5
+
+	addrIndexMap = make(map[string]uint64, 0)
+)
+
+// remainder stays on index 0
+type Faucet struct {
+	seed             *seed.Seed
+	remainderOutput  *faucetOutput
+	availableOutputs *queue.Queue[*faucetOutput]
+}
+
+// NewFaucet creates a new Faucet instance.
+func NewFaucet(seed *seed.Seed) *Faucet {
+	f := &Faucet{
+		seed:             seed,
+		availableOutputs: queue.New[*faucetOutput](numOfOutputsToSplit),
+	}
+
+	for i := uint64(0); i <= uint64(numOfOutputsToSplit); i++ {
+		addrIndexMap[f.seed.Address(i).Address().Base58()] = i
+	}
+
+	return f
+}
+
+// Start starts the faucet to fulfill faucet requests.
+func (f *Faucet) Start(ctx context.Context, requestChan <-chan *faucet.Payload) {
+	for {
+		select {
+		case p := <-requestChan:
+			blk, err := f.handleFaucetRequest(p)
+			if err != nil {
+				Plugin.LogErrorf("fail to send funds to %s: %v", p.Address().Base58(), err)
+				return
+			}
+			Plugin.LogInfof("send funds to %s: blockID: %s", p.Address().Base58(), blk.ID().Base58())
+
+		case <-ctx.Done():
+			return
+		}
+	}
+}
+
+// DeriveStateFromTangle derives unspent outputs of faucet from snapshot.
+func (f *Faucet) DeriveStateFromTangle() {
```

Comment:
This shouldn't be necessary if we use `client/wallet`, because it initializes the state automatically.